clang/AMDGPU: Set noalias.addrspace metadata on atomicrmw

arsenm opened the pull request.

arsenm committed:
clang/AMDGPU: Set noalias.addrspace metadata on atomicrmw

arsenm commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102462?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#102462** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102462?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#102461** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102461?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

shiltian commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -647,6 +647,14 @@ class LangOptions : public LangOptionsBase {
     return ConvergentFunctions;
   }
 
+  /// Return true if atomicrmw operations targeting allocations in private
```
Do we want to have a check in target machine to tell if atomic operation on specific address space is legal? I'm thinking of adding atomic support in `AAAddressSpace`, and could drop the address space if an atomic operation in an inferred address space is not legal.

arsenm replied:
This has nothing to do with the target, this is language semantics. What the target natively supports is an incomprehensibly complicated question, and is a legalization issue. We could annotate some operations with this in infer address spaces, but there's only a few cases where it's probably useful there

Artem-B commented on clang/lib/CodeGen/Targets/AMDGPU.cpp:
```diff
@@ -550,6 +551,16 @@ AMDGPUTargetCodeGenInfo::getLLVMSyncScopeID(const LangOptions &LangOpts,
+    llvm::MDNode *ASRange = MDHelper.createRange(
+        llvm::APInt(32, llvm::AMDGPUAS::PRIVATE_ADDRESS),
+        llvm::APInt(32, llvm::AMDGPUAS::PRIVATE_ADDRESS + 1));
+    RMW.setMetadata(llvm::LLVMContext::MD_noalias_addrspace, ASRange);
```
What's the plan for this metadata? It does not seem to exist or be used for anything yet. 
I do not see it in https://github.com/llvm/llvm-project/blob/6b5308b7924108d63149d7c521f21c5e90da7a09/llvm/include/llvm/IR/FixedMetadataKinds.def#L21

Does this patch miss some changes or depends on other patches?

arsenm replied:
It's defined in the parent PR #102461. I haven't posted the patch to use it yet, but the primary use is the AtomicExpand hooks

Artem-B commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -647,6 +647,14 @@ class LangOptions : public LangOptionsBase {
+  bool threadPrivateMemoryAtomicsAreUndefined() const {
+    // Should be false for OpenMP.
+    // TODO: Should this be true for SYCL?
+    return OpenCL || CUDA;
```
@gonzalobg -- Does NVIDIA define what happens if atomics are used on local address space?

@arsenm  atomics/and AS relationship seems to be a property of the target, not the language. I.e. we could potentially have a different answer for HIP on AMDGPU and CUDA on NVPTX, even though both would have `LangOpts.CUDA=true`.

arsenm replied:
This is entirely a language property. The target only comes in for the legalization decision in codegen

gonzalobg replied:
> @gonzalobg -- Does NVIDIA define what happens if atomics are used on local address space?

I agree with @arsenm that this is a language property. In CUDA C++, just like in C++, the behavior of atomics to automatic variables is well-defined, e.g., this is ok:

```
__device__ void foo() {
   cuda::atomic<...> x(0);
   x.fetch_add(1); // OK
}
```

When compiling to PTX, however, `atom` requires global or shared statespaces (or generic to those).
That is, for local memory, LLVM must generate code that does not use `atom`.
But that's a problem for the LLVM NVPTX backend to solve.

hvdijk commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -647,6 +647,14 @@ class LangOptions : public LangOptionsBase {
+  /// memory are undefined.
+  bool threadPrivateMemoryAtomicsAreUndefined() const {
+    // Should be false for OpenMP.
+    // TODO: Should this be true for SYCL?
```
I think this is meant to be true for SYCL in the future, but not just yet? [SYCL 2020 4.15.3](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references) specifies

> The `sycl::atomic_ref` class also has a template parameter `AddressSpace`, which allows the application to make an assertion about the address space of the object of type `T` that it references. The default value for this parameter is `access::address_space::generic_space`, which indicates that the object could be in either the global or local address spaces. If the application knows the address space, it can set this template parameter to either `access::address_space::global_space` or `access::address_space::local_space` as an assertion to the implementation. Specifying the address space via this template parameter may allow the implementation to perform certain optimizations. Specifying an address space that does not match the object’s actual address space results in undefined behavior.

It does not specifically call out the private address space as being undefined, but it says an address space that does not match what is specified is undefined, and provides no way to specify the private address space, so I think the end result is the same.

However, at the moment, the [legacy atomic types](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atom-types-depr) are also still available and the same logic cannot be applied to those, so barring any explicit statement that the private address space is undefined, I think it will be necessary to assume for now that this is well-defined.

arsenm replied:
This is now derived from the builtins rather than the language mode

arsenm commented:
ping

arsenm committed:
Change behavior per-builtin

Legacy atomic* apis are allowed to not work for private,
but not std::atomic style. Base this based on the language builtins,
rather than the language directly.

Also start handling cmpxchg.

arsenm commented:
ping
